plugins/radar-android-google-sleep/src/main/java/org.radarbase.passive.google.sleep/GoogleSleepManager.kt

```diff
@@ -0,0 +1,153 @@
+            val time = currentTime
+            val sleepStartTime: Double = sleepSegmentEvent.startTimeMillis / 1000.0
+            val sleepEndTime: Double = sleepSegmentEvent.endTimeMillis / 1000.0
+            val sleepDuration: Double = sleepSegmentEvent.segmentDurationMillis / 1000.0
+            val sleepStatus: SleepStatus = sleepSegmentEvent.status.toSleepStatus()
+
+            send(segmentEventTopic, GoogleSleepSegmentEvent(time, time, sleepStartTime,
+            sleepEndTime, sleepDuration, sleepStatus))
+        }
+    }
+
+     fun sendSleepSegmentData(sleepIntent: Intent) {
+        logger.info("Sleep classify event data received")
+        val sleepClassifyEvents: List<SleepClassifyEvent> = SleepClassifyEvent.extractEvents(sleepIntent)
+        sleepClassifyEvents.forEach {  sleepClassifyEvent ->
+            val time = sleepClassifyEvent.timestampMillis / 1000.0
+            val sleepConfidence = sleepClassifyEvent.confidence
+            val light = sleepClassifyEvent.light
+            val motion = sleepClassifyEvent.motion
+
+            send(classifyEventTopic, GoogleSleepClassifyEvent(time, currentTime, sleepConfidence, light, motion))
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun registerForSleepData() {
+        if ( isPermissionGranted() ) {
+            status = SourceStatusListener.Status.CONNECTING
+
+            ActivityRecognition.getClient(service)
+                .requestSleepSegmentUpdates(
+                    sleepPendingIntent,
+                    SleepSegmentRequest.getDefaultSleepSegmentRequest() )
+                .addOnSuccessListener {
+                    status = SourceStatusListener.Status.CONNECTED
+                    logger.info("Successfully subscribed to sleep data")
+                }
+                .addOnFailureListener {exception ->
+                    status = SourceStatusListener.Status.UNAVAILABLE
+                    logger.error("Exception while subscribing to sleep data: $exception")
+                }
+        }
+        else logger.warn("Permission not granted for ACTIVITY_RECOGNITION")
+    }
+
+    private fun isPermissionGranted(): Boolean {
+        val permission = if ( Build.VERSION.SDK_INT >= 29 ) Manifest.permission.ACTIVITY_RECOGNITION
+        else "com.google.android.gms.permission.ACTIVITY_RECOGNITION"
+        return ContextCompat.checkSelfPermission(service,permission) == PackageManager.PERMISSION_GRANTED
+    }
+
+    private fun createSleepPendingIntent() {
+        val intent = Intent(service, SleepReceiver::class.java)
+        sleepPendingIntent = PendingIntent.getBroadcast(service, SLEEP_DATA_REQUEST_CODE, intent,
+            PendingIntent.FLAG_CANCEL_CURRENT.toPendingIntentFlag(true))
+        logger.info("Sleep pending intent created")
+    }
+
+    private fun unRegisterFromSleepData() {
+        ActivityRecognition.getClient(service)
+            .removeSleepSegmentUpdates(sleepPendingIntent)
```

Comment:
You'll have to check that this intent even exists: `onClose` can be called before or while `start` is being called. Alternatively, make the `sleepPendingIntent: PendingIntent` `final` and assign it in the constructor `init`.